sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/TransactionalBulkExecutor.java

```diff
@@ -0,0 +1,472 @@
+            logger.debug(msg, args);
+        } else {
+            logger.warn(msg, args);
+        }
+    }
+
+    private void shutdown() {
+        if (this.isShutdown.compareAndSet(false, true)) {
+            logDebugOrWarning("Shutting down, Context: {}", this.operationContextText);
+            logger.debug("Shutdown complete, Context: {}", this.operationContextText);
+        }
+    }
+
+    public Flux<CosmosBulkOperationResponse<TContext>> execute() {
+        return this
+            .executeCore()
+            .doFinally((SignalType signal) -> {
+                if (signal == SignalType.ON_COMPLETE) {
+                    logDebugOrWarning("BulkExecutor.execute flux completed - # left items {}, Context: {}, {}",
+                        this.totalCount.get(),
+                        this.operationContextText,
+                        getThreadInfo());
+                } else {
+                    int itemsLeftSnapshot = this.totalCount.get();
+                    if (itemsLeftSnapshot > 0) {
+                        logInfoOrWarning("BulkExecutor.execute flux terminated - Signal: {} - # left items {}, Context: {}, {}",
+                            signal,
+                            itemsLeftSnapshot,
+                            this.operationContextText,
+                            getThreadInfo());
+                    } else {
+                        logDebugOrWarning("BulkExecutor.execute flux terminated - Signal: {} - # left items {}, Context: {}, {}",
+                            signal,
+                            itemsLeftSnapshot,
+                            this.operationContextText,
+                            getThreadInfo());
+                    }
+                }
+
+                this.dispose();
+            });
+    }
+
+    private Flux<CosmosBulkOperationResponse<TContext>> executeCore() {
+
+        // For transactional batches, batches are pre-constructed by the writer with proper partition key grouping.
+        // We just need to execute each pre-built batch.
+        Integer nullableMaxConcurrentCosmosPartitions = cosmosBulkExecutionOptions.getMaxConcurrentCosmosPartitions();
+        int maxConcurrentBatches = nullableMaxConcurrentCosmosPartitions != null ?
+            Math.max(256, nullableMaxConcurrentCosmosPartitions) : 256;
+
+        logDebugOrWarning("TransactionalBulkExecutor.executeCore with MaxConcurrentBatches: {}, Context: {}",
+            maxConcurrentBatches,
+            this.operationContextText);
+
+        return this.inputBatches
+            .publishOn(this.executionScheduler)
+            .onErrorMap(throwable -> {
+                logger.warn("{}: Error observed when processing input batches. Cause: {}, Context: {}",
+                    getThreadInfo(),
+                    throwable.getMessage(),
+                    this.operationContextText,
+                    throwable);
+                return throwable;
+            })
+            .flatMap(
```

Comment:
I think this scheduling is a bit too simple - the flatMap will allow a certainnumber of batches to be processed concurrently - independently whether they are spanning multiple physical partiitons or not. To optimize bulk jobs - ideally each Spark executor can saturate each physical partition - so like in BulkExecutor this probably needs grouping by physical partition - and then enforcing a certain limit of cocnurrent transactional batches per physical partition. This can also be done in a spearate PR - but I think it iwll some more thinking to get this right. Probably like dynamic batch size iin bulk we need a way to dynamically tune the allowed number of concurrent transactional batches per physical partition

Reply:
But this has to be addressed before letting any custoemr startuse this feature.